### ISSUE TITLE
Specification and workflow for the friend-to-friend API

### ISSUE THREAD

**Issue by zamentur** [YunoHost/issues#1597]
## Goal of the friend-to-friend self-hosting API
The goal of the friend-to-friend self-hosting API is to automate services providing between self-hosters and between self-hoster and some professional or non-profit trustable providers.

F2F API could be provided by a yunohost or not.

In the following specs, A is always the friend applicant and B the friend provider.

For example, if A need a borg backup repository, it can ask B, a friend provider, to open one. This can be made manually or this could be a bit automatize through the friend-to-friend self-hosting API.

### B authorizes A to make service request through the API
We don't want B to be spammed or trapped with tons of request for services. So the mechanism only works with prevalidated servers. It means at the first step the admin B, need to authorize the server A to open some services with some limits. Without this step A won't be able to ask for any services.

In yunohost it could be done thanks this command:
```
yunohost friend add DOMAIN_A.TLD SERVICE_TYPE [--args "LIMIT1=VALUE1&LIMIT2=VALUE2"] [ --alias DOMAIN_A_BIS.TLD IP_A...]
```
This command will use the x509 certificate of DOMAIN_A.TLD, if it's CA signed it will accept directly otherwise it will ask for approval of the admin.
This also activate the friend-to-friend API available on `https://DOMAIN_B.TLD/.well-known/self-hosting-federation/`

In our example of borg backup repo, it could be:
```
yunohost friend add DOMAIN_A.TLD borgbackup --args "quota=50G"
```

### A ask for a services to B
When needed, A could decide to activate a service on server B, for example to add the repo borg:
```
$ yunohost backup repository add DOMAIN_B.TLD -m borg --quota 40G --alert MAIL1 MAIL2 --alert-after 1d
passphrase:
```
- [ ] See if we can derivated from passphrase the passphrase for encryption and the password to manage the remote repo.

This command will then try to use the F2F API to activate the repository through the F2F API:

#### Request
```
POST https://DOMAIN_B.TLD/.well-known/self-hosting-federation/v1/service/borgbackup
{
   public_key: "ssh-ed25519 A[...]",
   quota: "40G",
   alert: "MAIL1, MAIL2",
   alert_delay: "1 day",
   password: "XXXXXXXX",
   origin: "domain_a.tld",
   origin_signature: "[...]"
}
Authenticate with the certificate of DOMAIN_A.TLD
```

#### Possible responses

201 Created / 202 Accepted (if service not really yet set up)
```
{
  id: "domain_a_tld",
   repository: "ssh://USER@DOMAIN.TLD:PORT/~/PATH"
}
```

401 Unauthorized (domain not added in friend authorization) / 403 Forbidden (added previously but the quota or the service not)
```
{
   code: 401,
   type: "Unauthorized",
   message: "Unauthorized"
}
```

### A could ask for updating params of the service needed

```
yunohost backup repository update DOMAIN_B.TLD -m borg --quota 50G --alert MAIL1 MAIL2 --alert-after 1d 
```
#### Request
```
PUT https://DOMAIN_B.TLD/.well-known/self-hosting-federation/v1/service/borgbackup/SERVICE_ID
{
   public_key: "ssh-ed25519 A[...]",
   quota: "50G",
   alert: "MAIL1, MAIL2",
   alert_delay: "1 day",
   password: "XXXXXXXX",
   origin: "domain_a.tld",
   origin_signature: "[...]"
}
Authenticate with the certificate of DOMAIN_A.TLD
```
#### Possible responses
```
200
{
  id: "domain_a_tld",
   repository: "ssh://USER@DOMAIN.TLD:PORT/~/PATH"
}
```

401 (if bad password) / 403 (if domain_a.tld has been added previously but the quota or the service not)
```
403
{
   code: 403,
   type: "Forbidden",
   message: "Quota too big retry with 30G"
}
```

**!! Important: with backup_mx, thanks to this mechanism we could update the list of authorized emails (to avoid spam)**

### A could delete the service

```
yunohost backup repository remove DOMAIN_B.TLD
```

#### Request
```
DELETE https://DOMAIN_B.TLD/.well-known/self-hosting-federation/v1/service/borgbackup/SERVICE_ID
{
   password: "XXXXXXXX",
   origin: "domain_a.tld",
   origin_signature: "[...]"
}
Authenticate with the certificate of DOMAIN_A.TLD
```
#### Possible responses
Correctly deleted
```
200 OK
{}
```

401 (if bad password or bad origin)
```
{
   code: 495,
   type: "SSL Certificate Error",
   message: "Unauthorized"
}
```

- [x] FIXME: If it's possible to delete the repo from the server A it could be a security risk if server A has been compromised cause the attacker could destroy the backup...

### Other API endpoints
It could be convenient to be able to list all authorizations and services already provided

## Services Authorization-Request-Response definition
Here is a non exhaustive list of services this API could offer. Process to add a kind of services should be defined.
TL;DR:
 - `borgbackup` service
 - `backup_mx` service
 - `smtp_relay` service
 - `ynh_monitor` service
 - `public_ip_discover` service
 - `reverse_proxy` service
 - `ldap` service ?
 - `sso` service ?

#### Authorization
```
quota: Quota needed
```

#### Request
```
ssh_public_key: Public key from client server
quota: Quota needed
alert: Alert mails (or sms?) if no change has been made on the repo
alert-after: Alert delay if no change has been made on the repo
```
- [ ] FIXME: what about if an attacker set the email of an other person ?

#### Response
```
id: ID of the service
repo: repo address to use
```

#### Authorization
```
retention: Retention in days
```
- [ ] FIXME: is it possible to set a quota ?

#### Request
```
domains: List of domains to relay
retention: Retention in days
accept: List of mail to accept
```
!! domains should be checked to be sure they are on same IP than the authenticated domains

#### Response
```
id: ID of the service
mx: Backup MX address to configure in DNS
```

#### Authorization
Rate limit ?

#### Request
```
domains: List of domains to relay
mails: List of mails to relay ?
```
#### Response
```
id: ID of the service
host: Relay host
port: Relay port
user: Relay user
password: Relay password
```

- [ ] FIXME: what about spf / dkim ?

#### Authorization
Rate limit ?
- [ ] TODO
#### Request
- [ ] TODO
#### Response
```
id: ID of the service
url: URL of remote diagnosis service
```

#### Authorization
N/A

#### Request
N/A

#### Response
```
id: ID of the service
url: URL of ip service
```

#### Authorization
- [ ] TODO
#### Request
- [ ] TODO
#### Response
- [ ] TODO

#### Authorization
- [ ] TODO
#### Request
- [ ] TODO
#### Response
- [ ] TODO

## Questions
- [ ] What should we do if origin change of main domain ?
- [ ] How to manage correctly instances with several domains / ips (ipv4/ipv6) ?

## Yunohost specific implementation
The F2F API could be provided by a moulinette API without root permission.
A systemd service triggered when we write in a protected file could be run and apply services configuration through the current yunohost API with root permission.

## IANA reservations
https://www.iana.org/assignments/well-known-uris/well-known-uris.xhtml

**Comment by zamentur** [comment]
To give some news. I am currently reading a lot about ActivityPub, ActivityStream, JSON-LD and webfinger.
Even if [i am a bit affraid about the way activitypub works and is used](https://forum.chatons.org/t/le-fediverse-un-espace-preserve-du-pistage/4319), i think we could be inspired by some mechanism.

## Webfinger (RFC 7033)
It's not directly ActivityPub but used to discover account in mastodon and other fediverse app. The Webfinger RFC allows to discover URI from `/.well-known/webfinger?resource=RESOURCE` endpoint. It means that the F2F API could be located on an other place than `https://example.org/.well-known/self-hosting-federation`.

Example:
```
$ curl https://example.org/.well-known/webfinger?resource=example.org&rel=http%3A%2F%2Fexample.org%2Frel%2Fself-hosting-federation
{
    "subject":"example.org",
    "aliases":["https://example.org"],
    "links":[{
         "rel":"https://example.org/self-hosting-federation",
         "type":"application/activity+json",
         "href":"https://example.org/yunohost/api/self-hosting-federation"
    }]
}
```
However:
- if this URL is already used by federated social media (or other app for example cups) we need to use some tricks to merge responses or balance the request (without breaking security against bruteforce)
- It's probably important to not list all other domain in this answer

## ActivityPub
Its built on the concept of inbox and outbox for each actor. Here our server can be our actor and having an inbox able to receive some push alert from a friend server could be very useful. The outbox could also be useful to allow to do configuration in the order that seems relevant at the moment.

But it's important to make the outbox readable only by allowed server and even, i think those server should be able to read only their messages and not the messages to other server (at the moment i am not sure it works really like this with ActivityPub)

## ActivityStream
Current ActivityStream doesn't contains Object and Activity we really need to make for example a backup repository provisionning or a monitoring device, however ActivityStream can be extended and it could be a good base to do this.

Thanks to ActivityStreams we could decide to share accepted emails with a secondary MX or to push remote diagnosis alert...
